### ISSUE TITLE
Implement access restrictions to `system.functions` table to prevent exposing UDF's CREATE statement

### ISSUE THREAD

**Issue by alexsubota** [ClickHouse/ClickHouse#41419]
version 22.8.4.7

How to reproduce:

1. add section to config.xml
`<access_control_improvements>
        <select_from_system_db_requires_grant>true</select_from_system_db_requires_grant>
</access_control_improvements>`
2. Execute `select * from system.functions`

As result you will get list of all functions and create queries of them. 
But if i do it right, i should get error message when i try to select tables from system schema (except some tables from https://github.com/ClickHouse/ClickHouse/pull/38970)

**Comment by alexey-milovidov** [comment]
Doesn't it work as it is supposed to?

**Comment by alexsubota** [comment]
> Doesn't it work as it is supposed to?

https://github.com/ClickHouse/ClickHouse/pull/38970
It says that only `tables`, `columns`, `databases`, and some constant tables like `one`, `contributors` have access. From which I conclude that the `functions `table should not be available.
If this table should be accessible, could you please suggest the best way at clickhouse to hide the decrypt key (key is used in the function). Dictionary and dictGet function?

**Comment by alexey-milovidov** [comment]
Do I understand correctly that you are creating a SQL UDF containing the encryption key and it should be hidden?
Something like this can help: https://github.com/ClickHouse/ClickHouse/issues/39809

**Comment by alexsubota** [comment]
> Do I understand correctly that you are creating a SQL UDF containing the encryption key and it should be hidden? Something like this can help: #39809

Yep, originally the question was about hiding the code of function, but hide encryption key from function will be great too

**Comment by Schum-io** [comment]
Are there any updates here? @alexey-milovidov 

ticket https://github.com/ClickHouse/ClickHouse/issues/39809 does not help in this case.

We have exactly the same case, we need a function that has a decryption key embedded in it, but we need to restrict access to viewing the source code or the key itself in this function.

Currently, it is possible to view the function's code
through `show functions`
or through `select create_query from system.functions` and there is no way to restrict access to viewing this metadata.

**Comment by alexey-milovidov** [comment]
No updates so far. Access to `system.functions` is unrestricted.
